amcl/src/amcl_node.cpp

```diff
@@ -1043,6 +1053,103 @@ AmclNode::globalLocalizationCallback(std_srvs::Empty::Request& req,
   return true;
 }
 
+/*
+ * 
+ * name: customPoseGenerator
+ * Sample from a custom set of particles equal to the Max Number of particles
+ * @geometry_msgs::PoseArray
+ * @pf_vector_t
+ * 
+ */
+pf_vector_t AmclNode::customPoseGenerator(void* arg)
+{
+  geometry_msgs::PoseArray* parray = (geometry_msgs::PoseArray*) arg;
+  geometry_msgs::Pose le = parray->poses.back();
+  tf2Scalar roll, pitch, yaw ;
+  
+  tf2::Quaternion quat(le.orientation.x,le.orientation.y,le.orientation.z,le.orientation.w);
+  tf2::Matrix3x3(quat).getRPY(roll, pitch, yaw);
+    
+  parray->poses.pop_back();
+  
+  pf_vector_t p;
+  p.v[0] = le.position.x;
+  p.v[1] = le.position.y;
+  p.v[2] = (double)yaw;//drand48() * 2 * M_PI - M_PI;
+  
+  return p;
+}
+
+bool AmclNode::setParticlesCallback(amcl::amcl_particles::Request& req,
+                                    amcl::amcl_particles::Response& res)
+{  
+  int no_of_particles = req.pose_array_msg.poses.size();
+  ROS_DEBUG_NAMED("amcl custom particles","Received set particles srv call, Pose Array Header seq = %i, stamp = %0.4f",
+                  req.pose_array_msg.header.seq,req.pose_array_msg.header.stamp.toSec());
+  if (no_of_particles == max_particles_)
```

Comment:
general style note: this code would have a lot less indentation if it were structured like:

```
if (no_of_particles != max_particles_)
{
  ROS_ERROR();
  res.success = false;
  return true;
}

// All the other code, without the extra level of indentation
```